lib/sqlalchemy/dialects/postgresql/json.py

```diff
@@ -258,47 +285,51 @@ class JSONB(JSON):
 
     __visit_name__ = "JSONB"
 
-    class Comparator(JSON.Comparator):
+    class Comparator(JSON.Comparator[_T]):
         """Define comparison operations for :class:`_types.JSON`."""
 
-        def has_key(self, other):
+        type: JSONB
+
+        def has_key(self, other: Any) -> ColumnElement[bool]:
             """Boolean expression.  Test for presence of a key (equivalent of
             the ``?`` operator).  Note that the key may be a SQLA expression.
             """
-            return self.operate(HAS_KEY, other, result_type=sqltypes.Boolean)
+            return self.operate(HAS_KEY, other, result_type=sqltypes.Boolean)  # type: ignore[return-value]  # noqa: E501
 
-        def has_all(self, other):
+        def has_all(self, other: Any) -> ColumnElement[bool]:
             """Boolean expression.  Test for presence of all keys in jsonb
             (equivalent of the ``?&`` operator)
             """
-            return self.operate(HAS_ALL, other, result_type=sqltypes.Boolean)
+            return self.operate(HAS_ALL, other, result_type=sqltypes.Boolean)  # type: ignore[return-value]  # noqa: E501
 
-        def has_any(self, other):
+        def has_any(self, other: Any) -> ColumnElement[bool]:
             """Boolean expression.  Test for presence of any key in jsonb
             (equivalent of the ``?|`` operator)
             """
-            return self.operate(HAS_ANY, other, result_type=sqltypes.Boolean)
+            return self.operate(HAS_ANY, other, result_type=sqltypes.Boolean)  # type: ignore[return-value]  # noqa: E501
 
-        def contains(self, other, **kwargs):
+        def contains(self, other: Any, **kwargs: Any) -> ColumnElement[bool]:
             """Boolean expression.  Test if keys (or array) are a superset
             of/contained the keys of the argument jsonb expression
             (equivalent of the ``@>`` operator).
 
             kwargs may be ignored by this operator but are required for API
             conformance.
             """
-            return self.operate(CONTAINS, other, result_type=sqltypes.Boolean)
+            return self.operate(CONTAINS, other, result_type=sqltypes.Boolean)  # type: ignore[return-value]  # noqa: E501
```

Comment:
let's create an override of `operate` that takes the generic from result_type when provided, so we avoid all these ignores

Reply:
Indeed; added, and it's much better with this. Thanks!